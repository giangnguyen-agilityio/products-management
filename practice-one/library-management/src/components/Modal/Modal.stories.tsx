import { type StoryFn, type Meta } from "@storybook/react";
import Modal from "./index";

export default {
  title: "Example/Modal",
  component: Modal,
  argTypes: {
    closeModal: { action: "Close modal button clicked" }
  },
} as Meta<typeof Modal>;

const Template: StoryFn<typeof Modal> = (args) => (
  <Modal {...args} />
);

export const Default = Template.bind({});
Default.args = {
    showModal: true,
    modalType: 'add'
};
