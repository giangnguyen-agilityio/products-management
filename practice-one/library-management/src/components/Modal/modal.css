.modal-overlay {
  width: 100%;
  height: 100%;
  background-color: var(--bg-overlay-color);
  position: absolute;
  padding: 10% 0;
  top: 0;
  right: 0;
}

.modal-container {
  width: 600px;
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  border: 2px solid var(--primary-color);
  border-radius: 20px;
  background-color: var(--secondary-text-color);
  overflow: hidden;
  padding: 20px;
}

.modal-delete {
  width: 500px;
}

.modal-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  text-transform: uppercase;
}

.modal-title {
  font-family: var(--tertiary-font);
  color: var(--primary-color);
  font-size: 22px;
}

.close-modal-btn {
  padding: 8px;
  border: 3px solid var(--secondary-color);
  background-color: unset;
  color: var(--secondary-color);
  font-family: var(--secondary-font);
  font-size: 18px;
  transition: all 0.3s ease-in-out;
  border-radius: 50%;
}

.close-modal-btn:hover {
  color: var(--primary-text-color);
  border: 3px solid var(--primary-text-color);
}

/* Media Query for Laptop Devices */
@media only screen and (max-width: 1024px) {
  .modal-container {
    width: 100%;
  }
}

/* Media Query for Mobile Devices */
@media only screen and (max-width: 480px) {
  .modal-overlay {
    padding: 0;
  }
}
