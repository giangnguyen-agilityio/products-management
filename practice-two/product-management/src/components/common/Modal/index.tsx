import React, { useRef } from 'react'
import {
  Modal as ChakraModal,
  ModalCloseButton,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalBody,
} from '@chakra-ui/react'
import { MODAL } from '@constants'
import Form from '@components/Form'
import { IProduct } from '@types'

// Interface for the props that the Modal component receives
interface ModalProps {
  id?: string
  isOpen: boolean
  closeModal: () => void
  modalType: MODAL.ADD | MODAL.EDIT
  productData?: IProduct
  onAdd?: (formData: IProduct) => Promise<void>
  onEdit?: (id: string, formData: IProduct) => Promise<void>
}

// The Modal component
const Modal: React.FC<ModalProps> = ({
  id,
  isOpen,
  productData,
  closeModal,
  modalType,
  onAdd,
  onEdit,
}) => {
  const btnRef = useRef(null)

  return (
    <ChakraModal
      finalFocusRef={btnRef}
      isOpen={isOpen}
      onClose={closeModal}
      scrollBehavior="inside"
      size="lg"
    >
      <ModalOverlay />

      <ModalContent background="background">
        <ModalHeader>
          {/* Close button */}
          <ModalCloseButton
            borderRadius="full"
            border="2px solid"
            borderColor="transparent"
            color="primary"
            _hover={{
              borderColor: 'primary',
            }}
          />
        </ModalHeader>

        <ModalBody>
          {/* Render the appropriate content based on the modalType */}
          {modalType === MODAL.ADD || modalType === MODAL.EDIT ? (
            <Form
              id={id}
              formType={modalType === MODAL.ADD ? MODAL.ADD : MODAL.EDIT}
              onAdd={onAdd}
              onEdit={onEdit}
              productData={productData}
            />
          ) : null}
        </ModalBody>
      </ModalContent>
    </ChakraModal>
  )
}

export default React.memo(Modal)
