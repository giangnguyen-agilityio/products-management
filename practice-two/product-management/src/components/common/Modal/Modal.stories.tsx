import type { Meta, StoryObj } from '@storybook/react'
import Modal from './index'
import { MODAL } from '@constants'

const meta = {
  title: 'Practice Two/Modal Component',
  component: Modal,
  argTypes: {
    onCloseModal: { action: 'Close the modal button clicked' },
    onAdd: { action: 'Add a new product button clicked' },
    onEdit: { action: 'Edit a product button clicked' },
  },
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Modal>

export default meta
type Story = StoryObj<typeof meta>

const productDataSample = {
  id: 'P01',
  name: 'iPhone 14 Pro Max',
  image: 'https://assets.nicepagecdn.com/d2cc3eaa/3159880/images/yttyy.jpg',
  discount: 11,
  oldPrice: 1263.8,
  newPrice: 1116.31,
  rate: 4.5,
  description:
    "The iPhone 14 Pro Max is Apple's flagship smartphone, featuring a Super Retina XDR display, A16 Bionic chip for powerful performance, and an advanced triple-camera system for exceptional photography. With premium design, Face ID, and MagSafe technology, it offers a top-tier user experience.",
}

export const ModalAdd: Story = {
  args: {
    isOpen: false,
    modalType: MODAL.ADD,
  },
}

export const ModalEdit: Story = {
  args: {
    isOpen: false,
    modalType: MODAL.EDIT,
    productData: productDataSample,
  },
}
