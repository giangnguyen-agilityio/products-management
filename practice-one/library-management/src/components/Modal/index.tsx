import React, { memo, useState } from 'react'

// Importing the Toast, Typography, Form, ConfirmModal and Button components
import Toast from '../Toast'
import Typography from '../Typography'
import Button from '../Button'
import Form from '../Form'
import ConfirmModal from '../ConfirmModal'

// Importing the Icon from the React-icons library
import { IoClose } from 'react-icons/io5'

// Importing the CSS file for styling
import './modal.css'

// Define the props for the Modal component
interface ModalProps {
  modalId: string
  closeModal: () => void
  showModal: boolean
  modalType: 'add' | 'edit' | 'delete'
}

const Modal: React.FC<ModalProps> = (props) => {
  const { modalId, modalType, showModal, closeModal } = props
  const [toastMessage, setToastMessage] = useState<string | null>(null)
  const [toastStatus, setToastStatus] = useState<'success' | 'failure' | null>(
    null
  )

  // Function to handle close the modal
  const handleCloseModal = (showToast = true): void => {
    closeModal()
    document.removeEventListener('keydown', handleKeyDown)
    // // Check if the toast should be shown
    if (!showToast) {
    // Set the toast message and status to null
      setToastMessage(null)
      setToastStatus(null)
    }
  }

  // Function to handle keydown events
  const handleKeyDown = (event: KeyboardEvent): void => {
    if (event.key === 'Escape') {
      handleCloseModal(false) // Pass false to indicate the ESC key is pressed
    }
  }

  // Function to handle button close modal clicked
  const handleCloseModalButtonClick = (): void => {
    handleCloseModal(false)
  }

  // Function to handle showing or hiding the toast
  const handleToast = (
    message: string,
    status: 'success' | 'failure'
  ): void => {
    setToastMessage(message)
    setToastStatus(status)
  }

  if (showModal) {
    document.addEventListener('keydown', handleKeyDown)
  }

  // Render the modal if showModal is true
  return showModal ? (
    <>
      <div className="modal-overlay">
        <div
          className={`modal-container ${
            modalType === 'delete' ? 'modal-delete' : ''
          }`}
        >
          <div className="modal-header">
            <Typography variant="h2" className="modal-title">
              {modalType === 'add'
                ? 'Add a new book'
                : modalType === 'edit'
                  ? 'Edit the book'
                  : null}
            </Typography>
            <Button
              size="large"
              variant="primary"
              className="close-modal-btn"
              aria-label="Close the modal button"
              onClick={handleCloseModalButtonClick}
            >
              <IoClose size={20} />
            </Button>
          </div>
          <div className="modal-content">
             {/* Render the appropriate content based on the modalType */}
            {modalType === 'add' || modalType === 'edit' ? (
              <Form
                id={modalId}
                formType={modalType === 'add' ? 'add' : 'edit'}
                onCloseModal={handleCloseModal}
                handleToast={handleToast}
              />
            ) : (
              <ConfirmModal
                id={modalId}
                text={'book'}
                onCloseModal={handleCloseModal}
                handleToast={handleToast}
              />
            )}
          </div>
        </div>
      </div>
    </>
  ) : (
    <>
      {/* Render the toast if toastMessage and toastStatus are not null */}
      {toastMessage != null && toastStatus != null && (
        <Toast message={toastMessage} status={toastStatus} duration={5000} />
      )}
    </>
  )
}

export default memo(Modal)
