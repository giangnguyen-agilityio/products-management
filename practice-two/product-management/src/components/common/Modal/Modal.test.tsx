import { render, screen, fireEvent } from '@testing-library/react'
import Modal from './'
import { MODAL } from '@constants'

// Mock the dependencies and props
jest.mock('@components/Form', () => {
  return function MockedForm() {
    return <div data-testid="mocked-form"></div>
  }
})

const mockCloseModal = jest.fn()
const mockHandleAdd = jest.fn()
const mockHandleEdit = jest.fn()

const defaultProps = {
  isOpen: true,
  closeModal: mockCloseModal,
  modalType: MODAL.ADD,
  handleAdd: mockHandleAdd,
  handleEdit: mockHandleEdit,
}

describe('Modal Component', () => {
  it('renders without crashing', () => {
    const { container } = render(<Modal {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  it('calls closeModal when close button is clicked', () => {
    render(<Modal {...defaultProps} />)
    const closeButton = screen.getByLabelText('Close')
    fireEvent.click(closeButton)
    expect(mockCloseModal).toHaveBeenCalledTimes(1)
  })

  it('renders the form when modalType is add', () => {
    render(<Modal {...defaultProps} />)
    const form = screen.getByTestId('mocked-form')
    // Assertions
    expect(form).toBeInTheDocument()
  })

  it('renders the form when modalType is edit', () => {
    const props = { ...defaultProps, modalType: MODAL.EDIT }
    render(<Modal {...props} />)
    const form = screen.getByTestId('mocked-form')
    expect(form).toBeInTheDocument()
  })

  it('does not render the form when modalType is delete', () => {
    const props = { ...defaultProps, modalType: MODAL.DELETE }
    render(<Modal {...props} />)
    const form = screen.queryByTestId('mocked-form')
    expect(form).toBeNull()
  })
})
