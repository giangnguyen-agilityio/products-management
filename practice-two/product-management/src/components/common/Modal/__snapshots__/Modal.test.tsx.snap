// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component renders without crashing 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
